### PR TITLE
My go at the PinPoint challenges.

### PR DESCRIPTION
**Pull request by GeAR522**
This is my go at the PinPoint challenges.

I've commented most of what I was doing in each page and commented out stuff from the original if its not used anymore. 
Hope this is statisfactory for now and I'm happy to improve and learn whatever else is needed.

### CODE REVIEW

File: app/javascript/components/challenges/Counter.tsx [hunk 1]
```diff
@@ -3,23 +3,40 @@ import * as React from 'react';
...
 
       <div>
         <input readOnly value={count} />
       </div>
       <div>
-        <button onClick={() => setCount(count ** 2)}>Decrease</button>
+        <button onClick={() => setCount(count - 1)}>Decrease</button>
         <button onClick={() => setCount(count + 1)}>Increase</button>
+        <button onClick={() => setCount(reset)}>Reset</button>
+        <button onClick={() => setReset(count)}>Set Current as Reset Value</button>
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/Counter.tsx:25]: All looks good. Minor point but ideally you shouldn't use inline anonymous functions i.e. these should be defined above and then used e.g.
```
const handleDecrease = () => setCount(count - 1)

<button onClick={handleDecrease}>Decrease</button>
```

It just stops the function being redefined on each re-render

File: app/javascript/components/challenges/NextButton.tsx [hunk 2]
```diff
@@ -3,9 +3,9 @@ import * as React from 'react';
 export default function NextButton() {
   const [left, setLeft] = React.useState<number>(0);
   const [top, setTop] = React.useState<number>(0);
-
+// made the max of random 0, so it doesn't move after the first translation.
   function handleMouseMove(event: React.MouseEvent<HTMLDivElement>) {
-    const randomIndex = Math.floor(Math.random() * 4);
+    const randomIndex = Math.floor(Math.random() * 0);
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/NextButton.tsx:8]: Fair enough. Another solution would be to remove the `onMouseMove`

File: app/javascript/components/challenges/Login.tsx [hunk 3]
```diff
@@ -19,6 +25,22 @@ export default function Login() {
 
   const authToken = document.querySelector('head meta[name="csrf-token"]' as any).content;
 
+  // toggle to show/hide password
+  const [showP, setShowP] = React.useState<boolean>(false);
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/Login.tsx:29]: Again a minor point but don't abbreviate variables makes it more difficult for the next person

File: app/javascript/components/challenges/Login.tsx [hunk 4]
```diff
@@ -29,12 +51,12 @@ export default function Login() {
...
-        <input name="email" type="text" value={email} />
-        <div style={{ color: 'red', margin: '10px 0' }}>{/* Email validation errors go here */}</div>
+        <input name="email" type="email" onChange={onChange} value={email} onBlur={handleBlur} />
+        {!valid && <div style={{ color: 'red', margin: '10px 0' }}>E.g. something@somewhere.com</div>}
         <label htmlFor="">Password</label>
         <div style={{ display: 'flex', marginBottom: '20px' }}>
-          <input name="password" type="password" />
-          <button type="button">Show Password</button>
+          <input name="password" type={!showP ? "password" : "text"} />
+          <button type="button" onClick={() => setShowP(!showP)}>Show Password</button>
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/Login.tsx:59]: As per my other comment don't have the `onClick` function inline. Also would be nice to change the button text in line with whether we are showing or hiding e.g.

```
{showP ? 'Hide' : 'Show'} Password
```

File: app/javascript/components/challenges/Login.tsx [hunk 5]
```diff
@@ -19,6 +25,22 @@ export default function Login() {
...
+  const [showP, setShowP] = React.useState<boolean>(false);
+
+  
+  // changing the input type to email will validate the format, but doesn't work onblur,
+  // so just doing it manually.
+    
+  const [valid, setValid] = React.useState<boolean>(true);
+  // every blur will check the email, so starting the boolean as true to hide the error message when first loading.
+  const handleBlur = () => {
+    const emailCheck = /(.+@.+)((\.com)|(\.co\.uk)|(\.co\.je)|(\.me))/i.test(email)
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/Login.tsx:38]: Where did you get the regex from out of interest?

File: app/javascript/components/challenges/Cards.tsx [hunk 6]
```diff
@@ -4,29 +4,84 @@ export default function Cards() {
...
+// to create the pagination buttons I used for loop to fill an empty array with
+// the html outputs to be rendered. then made them into buttons to match the previous
+// buttons and copied the style.
+
+  function Pagination() {
+    const pages = [];
+    for (let i = 0; i < pictureDates.length; i++) {
+      pages.push(<button className="page-item" style={buttonStyles} onClick={() => setPicIndex(i)}>{i + 1}</button>);
+    }
+    return <>{pages}</>;
```
Comment (BowlegsBill) [inline app/javascript/components/challenges/Cards.tsx:80]: Would be simpler to do 
```
cosnt paginationButtons = pictureDates.map((date) => {
  return <button className="page-item" style={buttonStyles} onClick={() => setPicIndex(i)}>{i + 1}</button>
})
```

And then just click that array into the return

```
<button style={buttonStyles} onClick={minusPicIndex}>Previous</button>
{paginationButtons}
<button style={buttonStyles} onClick={plusPicIndex}>Next</button>
```

Comment (BowlegsBill) [reply app/javascript/components/challenges/Cards.tsx:80]: Or you could do it all inline:

```
<button style={buttonStyles} onClick={minusPicIndex}>Previous</button>
{pictureDates.map((date) => {
  return <button className="page-item" style={buttonStyles} onClick={() => setPicIndex(i)}>{i + 1}</button>
})}
<button style={buttonStyles} onClick={plusPicIndex}>Next</button>
```

**Review by BowlegsBill — APPROVED** [review]
Really impressed given you haven't really looked at React previously 👍 

I've added a couple of suggestions to give you an idea of some of our coding standards

File: app/javascript/components/challenges/Login.tsx (hunk 5 above)
Comment (GeAR522) [reply app/javascript/components/challenges/Login.tsx:38]: I first learnt Regex when I was learning Python. I had to go double check the differences to write it in JS but I use this site to check my regex before adding it in: https://regexr.com/

**Pull request closed without merging**